File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.js
function solution(s) {
    var answer = 0;
    let s_lst = Array.from(s);
    let len_s = s.length;
    const map = new Map()
    map.set(')', '(')
    map.set('}', '{')
    map.set(']', '[')
    for (let start = 0; start < len_s; start++) {
        let stk = [];
        let isO = true;
        for (let i = 0; i < len_s; i++) {
            if (s[(start+i)%len_s] === '(' || s[(start+i)%len_s] === '{' || s[(start+i)%len_s] === '[') {
                stk.push(s[(start+i)%len_s])
            } else {
                const pair = map.get(s[(start+i)%len_s])
                if (pair === stk[stk.length - 1]) {
                    stk.pop()
                } else {
                    isO= false;
                    break
                }
            }
            
        } 
        if (!stk.length && isO) {
            answer ++;
        }
        
    }
    
    
    return answer;
}

File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.java
import java.util.*;

class Solution {
    static char[] sArr;
    static Character[] startsList = {'(', '[', '{'};
    
    static boolean isRight(int stIdx) {
        Stack<Character> st = new Stack<>();
        List<Character> starts = new ArrayList<>(Arrays.asList(startsList));
        
        for (int d=0; d<sArr.length; d++) {
            char c = sArr[(stIdx+d)%sArr.length];
            if (starts.contains(c)) {
                st.push(c);
                continue;
            } 
            if (c == ')') {
                if (st.size() > 0 && st.peek() == '(') {
                    st.pop();
                    continue;
                } else {
                    return false;
                }
            }
            if (c == ']') {
                if (st.size() > 0 && st.peek() == '[') {
                    st.pop();
                    continue;
                } else {
                    return false;
                }
            }
            if (c == '}') {
                if (st.size() > 0 && st.peek() == '{') {
                    st.pop();
                    continue;
                } else {
                    return false;
                }
            }
        }
        if (st.size() == 0) return true;
        return false;
    }
    
    public int solution(String s) {
        int answer = 0;
        
        sArr = s.toCharArray();
        for (int st=0; st<sArr.length; st++) {
            if (isRight(st)) answer++;
        }
        
        return answer;
    }
}